// discord.js-commando/src/commands/util/prefix.js
const {stripIndents, oneLine} = require('common-tags');
const Command = require('../base');
const {guild, keys} = require('../../../../../db/JSONSListeners');
const {updatePrefix} = require('../../../../../db/DBUpdateGuildSetting');

module.exports = class PrefixCommand extends Command {
    constructor(client) {
        super(client, {
            name: 'prefix',
            group: 'util',
            memberName: 'prefix',
            description: 'Muestra o configura el prefijo para los comandos.',
            format: '[prefix/"default"/"none"]',
            details: oneLine`
				Si no se entrega prefijo, el actual será mostrado.
				Si el prefijo es "default", el prefijo será reiniciado al prefijo original del bot.
				Solo los administradores pueden cambiar el prefijo.
			`,
            examples: ['prefix', 'prefix -', 'prefix omg!', 'prefix default'],

            args: [
                {
                    key: 'prefix',
                    prompt: '¿Cuál te gustaría que fuese el prefijo del bot?',
                    type: 'string',
                    max: 15,
                    default: ''
                }
            ]
        });
    }

    async run(msg, args) {
        // Just output the prefix
        if (!args.prefix) {
            const guild_data = await guild(msg.guild.id);
            const prefix = guild_data[keys.prefix];

            return msg.reply(stripIndents`
				${prefix ? `El prefijo de los comandos es \`\`${prefix}\`\`.` : 'No hay prefijo de comandos.'}
			`);
        }

        // Check the user's permission before changing anything
        if (msg.guild) {
            if (!msg.member.hasPermission('ADMINISTRATOR') && !this.client.isOwner(msg.author)) {
                return msg.reply('Solo los administradores puede cambiar el prefijo.');
            }
        } else if (!this.client.isOwner(msg.author)) {
            return msg.reply('Solo el dueño del bot puede cambiar el prefijo por defecto.');
        }

        // Save the prefix
        const lowercase = args.prefix.toLowerCase();
        const prefix = lowercase === 'none' ? '' : args.prefix;
        if (prefix === '') {
            msg.reply("No puedo dejar el prefijo en blanco");
        }
        else if (lowercase === 'default') {
            updatePrefix(msg.guild.id, "~")
                .then(
                    msg.reply("La configuración ha sido actualizada con exito.")
                )
                .catch(() =>
                    msg.reply("Por alguna razón, no he podido actualizar las configuraciones, " +
                        "si este problema persiste, informale al desarrollador a traves del comandos `feedback`")
                );
        } else {
            updatePrefix(msg.guild.id, lowercase)
                .then(
                    msg.reply("La configuración ha sido actualizada con exito.")
                )
                .catch(() =>
                    msg.reply("Por alguna razón, no he podido actualizar las configuraciones, " +
                        "si este problema persiste, informale al desarrollador a traves del comandos `feedback`")
                );
        }
        return null;
    }
};
