// discord.js-commando/src/commands/util/prefix.js
const { stripIndents, oneLine } = require('common-tags');
const Command = require('../base');

module.exports = class PrefixCommand extends Command {
	constructor(client) {
		super(client, {
			name: 'prefix',
			group: 'util',
			memberName: 'prefix',
			description: 'Mostrar o configurar el prefijo de comandos.',
			format: '[prefix/"default"/"none"]',
			details: oneLine`
				Si no se proporciona un prefijo, se mostrar치 el prefijo actual.
				Si el prefijo es "predeterminado", el prefijo se restablecer치 al prefijo predeterminado del bot.
				Si el prefijo es "ninguno", el prefijo se eliminar치 por completo, solo permitiendo que las menciones ejecuten comandos.
				Solo los administradores pueden cambiar el prefijo.
			`,
			examples: ['prefix', 'prefix -', 'prefix omg!', 'prefix default', 'prefix none'],

			args: [
				{
					key: 'prefix',
					prompt: 'What would you like to set the bot\'s prefix to?',
					type: 'string',
					max: 15,
					default: ''
				}
			]
		});
	}

	async run(msg, args) {
		// Just output the prefix
		if(!args.prefix) {
			const prefix = msg.guild ? msg.guild.commandPrefix : this.client.commandPrefix;
			return msg.reply(stripIndents`
				${prefix ? `The command prefix is \`${prefix}\`.` : 'There is no command prefix.'}
				To run commands, use ${msg.anyUsage('command')}.
			`);
		}

		// Check the user's permission before changing anything
		if(msg.guild) {
			if(!msg.member.hasPermission('ADMINISTRATOR') && !this.client.isOwner(msg.author)) {
				return msg.reply('Solos los administradores pueden cambiar el prefijo de comandos.');
			}
		} else if(!this.client.isOwner(msg.author)) {
			return msg.reply('Solo el due침o del bot puede administrar los prefijos globales.');
		}

		// Save the prefix
		const lowercase = args.prefix.toLowerCase();
		const prefix = lowercase === 'none' ? '' : args.prefix;
		let response;
		if(lowercase === 'default') {
			if(msg.guild) msg.guild.commandPrefix = null; else this.client.commandPrefix = null;
			const current = this.client.commandPrefix ? `\`${this.client.commandPrefix}\`` : 'no prefix';
			response = `Reset the command prefix to the default (currently ${current}).`;
		} else {
			if(msg.guild) msg.guild.commandPrefix = prefix; else this.client.commandPrefix = prefix;
			response = prefix ? `Set the command prefix to \`${args.prefix}\`.` : 'Removed the command prefix entirely.';
		}

		msg.reply(`${response} To run commands, use ${msg.anyUsage('command')}.`);
		return null;
	}
};
