// discord.js-commando/src/util.js
function escapeRegex(str) {
	return str.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&');
}

function disambiguation(items, label, property = 'name') {
	const itemList = items.map(item => `"${(property ? item[property] : item).replace(/ /g, '\xa0')}"`).join(',   ');
	return `Multiple ${label} found, please be more specific: ${itemList}`;
}

function paginate(items, page = 1, pageLength = 10) {
	const maxPage = Math.ceil(items.length / pageLength);
	if(page < 1) page = 1;
	if(page > maxPage) page = maxPage;
	const startIndex = (page - 1) * pageLength;
	return {
		items: items.length > pageLength ? items.slice(startIndex, startIndex + pageLength) : items,
		page,
		maxPage,
		pageLength
	};
}

const permissions = {
	ADMINISTRATOR: 'Administrador',
	VIEW_AUDIT_LOG: 'Ver registro de auditoría',
	MANAGE_GUILD: 'Gestionar servidor',
	MANAGE_ROLES: 'Gestionar roles',
	MANAGE_CHANNELS: 'Gestionar canales',
	KICK_MEMBERS: 'Expulsar miembros',
	BAN_MEMBERS: 'Banear miembros',
	CREATE_INSTANT_INVITE: 'Crear invitación',
	CHANGE_NICKNAME: 'Cambiar apodo',
	MANAGE_NICKNAMES: 'Gestionar apodos',
	MANAGE_EMOJIS: 'Gestionar emojis',
	MANAGE_WEBHOOKS: 'Gestionar webhooks',
	VIEW_CHANNEL: 'Leer canales de texto y ver canales de voz',
	SEND_MESSAGES: 'Enviar mensajes',
	SEND_TTS_MESSAGES: 'Enviar mensajes de texto a voz',
	MANAGE_MESSAGES: 'Gestionar mensajes',
	EMBED_LINKS: 'Insertar enlaces',
	ATTACH_FILES: 'Adjuntar archivos',
	READ_MESSAGE_HISTORY: 'Leer el historial de mensajes',
	MENTION_EVERYONE: 'Mencionar @everywhere, @here y todos los roles',
	USE_EXTERNAL_EMOJIS: 'Usar emojis externos',
	ADD_REACTIONS: 'Añadir reacciones',
	CONNECT: 'Conectar',
	SPEAK: 'Hablar',
	MUTE_MEMBERS: 'Silenciar miembros',
	DEAFEN_MEMBERS: 'Ensordecer miembros',
	MOVE_MEMBERS: 'Mover miembros',
	USE_VAD: 'Usar actividad de voz'
};

module.exports = {
	escapeRegex,
	disambiguation,
	paginate,
	permissions
};
